docs/cheat-sheets/python-command-injection.mdx

```diff
@@ -0,0 +1,649 @@
+`eval()` can be dangerous if used to evaluate dynamic content.
+If this content can be input from outside the program, this may be a code injection vulnerability.
+
+Example:
+
+```python
+# Value supplied by user
+user_input = "__import__('code').InteractiveInterpreter().runsource('import requests;requests.get(\'localhost:3000\')')"
+
+# Vulnerable
+eval(user_input)
+```
+
+#### References:
+
+- [**eval()** documentation](https://docs.python.org/3/library/functions.html#eval)
+
+#### Mitigation
+
+Do not use `eval()`. Alternatively,
+
+- Ensure evaluated content is not definable by external sources.
+- If it is not possible - strip everything except alphanumeric characters from an input provided for the command string and arguments.
+- Don’t try to make `eval` safe with tricks like `{'__builtins__':{}}`
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.eval-detected.eval-detected" />
+
+### 2.C. Accepting logging configuration with **logging.config.listen()**
+
+`logging.config.listen()` starts up a socket server on the specified port, and listens for new configurations.
+Because portions of the logging configuration are passed through `eval()`, use of this function may open its users to a security risk.
+While the function only binds to a socket on localhost, and so does not accept connections from remote machines, there are scenarios where untrusted code could be run under the account of the process which calls `listen()`.
+
+Example:
+
+```python
+# Server example: starting up a socket server on 9999 port, and listening for new configurations.
+import logging
+import logging.config
+
+logging.config.fileConfig('logging.conf')
+t = logging.config.listen(9999)
+t.start()
+
+
+# Client example: sending configuration from `data_to_send` variable to localhost:9999
+import socket, sys, struct
+
+# Config example: print("pwned") will be evaluated and "pwned" will be printed to the console
+data_to_send = """
+[loggers]
+keys=root
+
+[handlers]
+keys=hand01
+
+[formatters]
+keys=form01
+
+[logger_root]
+level=NOTSET
+handlers=hand01
+
+[handler_hand01]
+class=StreamHandler
+level=NOTSET
+formatter=form01
+args=(print("pwned"),)
+
+[formatter_form01]
+format=F1 %(asctime)s %(levelname)s %(message)s
+datefmt=
+class=logging.Formatter
+"""
+
+s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+s.connect(('localhost', 9999))
+s.send(struct.pack('>L', len(data_to_send)))
+s.send(data_to_send)
+s.close()
+```
+
+#### References:
+
+- [**logging.config.listen()** documentation](https://docs.python.org/3/library/logging.config.html#logging.config.listen)
+
+#### Mitigation
+
+Verify what is sent across the socket. Alternatively,
+
+- To avoid the risk, use the verify argument to `logging.config.listen()` to prevent unrecognised configurations from being applied. This could be done by encrypting and/or signing what is sent across the socket, such that the verify callable can perform signature verification and/or decryption.
```

Comment:
```suggestion
- To avoid the risk, use the verify argument to `logging.config.listen()` to prevent unrecognized configurations from being applied. This could be done by encrypting and/or signing what is sent across the socket, such that the verify callable can perform signature verification and/or decryption.
```